### ISSUE TITLE
hdf5 1.10.x mpi hang issue

### ISSUE THREAD

**Issue by jedwards4b** [Unidata/netcdf-c#1120]
> Please provide as much of the following information as you can, as applicable to the issue being reported.  Naturally, not all information is relevant to every issue, but the more information we have to start, the better!

### Environment Information
The problem is in the 1.10.x versions of the HDF5 library.   It has been reported to the HDF5 developers 
as issue HDFFV-10501 - however their response seems to be to refuse to address the issue.   

> Feel free to skip this if the issue is related to documentation, a feature request, or general discussion.

* What platform are you using? (please provide specific distribution/version in summary)
  * [ x] Linux
  * [ ] Windows
  * [ ] OSX
  * [ ] Other
  * [ ] NA
* 32 and/or 64 bit?
  * [ ] 32-bit
  * [x ] 64-bit
* What build system are you using?
  * [x ] autotools (`configure`)
  * [ ] cmake
* Can you provide a sample netCDF file or `C` code to recreate the issue?
  * [x ] Yes (please attach to this issue, thank you!)
  * [ ] No
  * [ ] Not at this time

### Summary of Issue
When a distributed variable is written with one decomposition strategy and then read with another (after a call to nc_sync) the read hangs in MPI.   

### Steps to reproduce the behavior
Run the following on four mpi tasks
```
#include <netcdf.h>
#include <stdio.h>
#include <stdlib.h>
#include <netcdf_par.h>
#include <mpi.h>

#define ERR(i) fprintf(stderr,"ERROR rank %d line %d ierr = %d\n",rank,__LINE__, i)
#define NDIMS 4
#define VAR_NAME "var1"
#define DIM1_NAME "Time"
#define DIM2_NAME "X"
#define DIM3_NAME "Y"
#define DIM4_NAME "Z"

void * Malloc (size_t n)
{
    void *ptr;
    ptr = malloc (n);
    if (!ptr)
    {
	fprintf (stderr, " could not allocate %lu bytes\n", n);
	exit (EXIT_FAILURE);
    }
    return ptr;
}


int
main (int argc, char **argv)
{
    int rank, nproc, ierr, i, j, k, l, m;
    const char *name = "try.nc";
    char new_name[32];
    char mpi_name[128];
    int namelen = 127;
    int *result;
    size_t *count;
    size_t *start;
    int *dimids;
    int varid;
    int ncid;
    int count_total;
    int ndims_in, nvars_in, natts_in, unlimdimid_in;
    int *result_in;
    int *tmp_unlimdimids, tmp_nunlimdims;
    MPI_Comm newcomm;

    MPI_Init (&argc, &argv);
    MPI_Comm_size (MPI_COMM_WORLD, &nproc);
    MPI_Comm_rank (MPI_COMM_WORLD, &rank);
    MPI_Get_processor_name (mpi_name, &namelen);
    MPI_Comm_dup(MPI_COMM_WORLD, &newcomm);

    ierr = nc_set_log_level(5);

    count_total = 40;
    result = NULL;
    count = (size_t *) Malloc (sizeof (int) * NDIMS);
    start = (size_t *) Malloc (sizeof (int) * NDIMS);
    dimids = (int *) Malloc (sizeof (int) * NDIMS);
    result_in = NULL;

    if (rank == 0){
	count[0] = 1;
	count[1] = 1;
	count[2] = 10;
	count[3] = 1;
	start[0] = 0;
	start[1] = 0;
	start[2] = 0;
	start[3] = 0;
	count_total = 10;
	result = (int *) Malloc (sizeof (int) * count_total);
    }else if (rank == 1){
	count[0] = 1;
	count[1] = 1;
	count[2] = 10;
	count[3] = 1;
	start[0] = 0;
	start[1] = 1;
	start[2] = 0;
	start[3] = 0;
	count_total = 10;
	result = (int *) Malloc (sizeof (int) * count_total);
    }else if (rank == 2){
	count[0] = 1;
	count[1] = 1;
	count[2] = 10;
	count[3] = 1;
	start[0] = 0;
	start[1] = 0;
	start[2] = 0;
	start[3] = 1;
	count_total = 10;
	result = (int *) Malloc (sizeof (int) * count_total);
    }else if (rank == 3){
	count[0] = 1;
	count[1] = 1;
	count[2] = 10;
	count[3] = 1;
	start[0] = 0;
	start[1] = 1;
	start[2] = 0;
	start[3] = 1;
	count_total = 10;
	result = (int *) Malloc (sizeof (int) * count_total);
    }

    for (i = 0; i < count_total; ++i)
	result[i] = 100*rank + i;

  fprintf(stderr,"rank %d creating %s on %s\n",rank,name,mpi_name);

  if (ierr = nc_create_par (name, NC_MPIIO|NC_NETCDF4, newcomm, MPI_INFO_NULL, &ncid))
      ERR(ierr);
  ierr = nc_set_fill(ncid, NC_NOFILL, NULL);
  fprintf(stderr,"ncid= %d\n",ncid);
  /* Create three dimensions. */
  if (ierr = nc_def_dim (ncid, DIM4_NAME, 2, &dimids[3]))
      ERR(ierr);
  if (ierr = nc_def_dim (ncid, DIM3_NAME, 10, &dimids[2]))
      ERR(ierr);
  if (ierr = nc_def_dim (ncid, DIM2_NAME, 2, &dimids[1]))
      ERR(ierr);
  if (ierr = nc_def_dim (ncid, DIM1_NAME, NC_UNLIMITED, &dimids[0]))
      ERR(ierr);

  /* Create one var. */
  if (ierr = nc_def_var (ncid, VAR_NAME, NC_INT, NDIMS, dimids, &varid))
      ERR(ierr);

  /* Write metadata to file. */
  if (ierr = nc_enddef (ncid))
      ERR(ierr);

  ierr = nc_inq_unlimdims(ncid, &tmp_nunlimdims, NULL);

  if (ierr = nc_var_par_access (ncid, varid, NC_COLLECTIVE))
      ERR(ierr);

  if (ierr = nc_put_vara_int(ncid,varid,start,count,result)) ERR(ierr);

  /* write a second unlimdim */
  start[0] = 1;
  if (ierr = nc_put_vara_int(ncid,varid,start,count,result)) ERR(ierr);

/*  MPI_Barrier(newcomm); */

  if ( ierr = nc_sync(ncid)) ERR(ierr);

  if (result != NULL) free(result);

  start[0] = 0;
  start[1] = 0;
  start[2] = 0;
  start[3] = 0;
  if (rank == 1)
  {
      count[0] = 1;
      count[1] = 2;
      count[2] = 10;
      count[3] = 2;
      count_total = 40;
      result_in = (int *)Malloc(sizeof(int)*count_total);
  }else{
      count[0] = 0;
      count[1] = 0;
      count[2] = 0;
      count[3] = 0;
      count_total = 0;
  }

  /* Read one slab of data. */
  fprintf(stderr,"%d reading in %d variables\n",rank,count_total);
  if (ierr = nc_get_vara_int (ncid, varid, start, count, result_in))
      ERR(ierr);
  if (rank == 1)
  {
      i = 0;
      for (j = 0; j < count[0]; j++)
      {
          for (k = 0; k < count[1]; k++)
	  {
              for (l = 0; l < count[2]; l++)
	      {
		  for (m = 0; m < count[3]; m++, i++)
		  {
		      fprintf (stderr, "%d %d %d %d (%d) - %d \n", j, k, l, m, i, result_in[i]);
		  }
	      }
	  }
      }
      free(result_in);
  }

  /* Close the netcdf file. */
  if (ierr = nc_close (ncid))
      ERR(ierr);

  /* should free data put leave it for later */

  MPI_Finalize ();

  return EXIT_SUCCESS;
}
```

**Comment by edhartnett** [comment]
Thanks I will make this a test.

This is the same as issue #1016. Thanks for isolating the problem into a test.

**Comment by edwardhartnett** [comment]
This issue has been fixed in HDF5 starting with 1.10.4. This issue may be closed.
